lib/msprime.c

```diff
@@ -2346,17 +2371,38 @@ msp_dtwf_generation(msp_t *self)
 
 /* The main event loop for the Wright Fisher model. */
 static int WARN_UNUSED
-msp_run_dtwf(msp_t *self, double max_time)
+msp_run_dtwf(msp_t *self, double max_time, unsigned long max_events)
 {
     int ret = 0;
-
-    while (msp_get_num_ancestors(self) > 0 && self->time < max_time) {
+    size_t transition_pop_size = 0;
+    double num_dtwf_generations;
+    
+    num_dtwf_generations = self->model.params.dtwf.num_dtwf_generations;
+
+    /* printf("Running %f dtwf generations...\n", num_dtwf_generations); */
+    /* printf("Starting with %lu ancestors\n", msp_get_num_ancestors(self)); */
+    while (msp_get_num_ancestors(self) > 0) {
+        if (self->time == num_dtwf_generations) {
+            break;
+        }
+        /* printf("%f\n", self->time); */
         self->time++;
         ret = msp_dtwf_generation(self);
         if (ret != 0) {
             goto out;
         }
     }
+    /* printf("dtwf simulations complete\n"); */
+    /* printf("%lu ancestors remaining\n", msp_get_num_ancestors(self)); */
+
+    transition_pop_size = 1; //TODO: Verify correct value
```

Comment:
This is a bit awkward all right. I think it's probably safer to call run(time) at the top level and then change the model in the simulator. There'll be all sorts of odd side effects otherwise I think.